### ISSUE TITLE
Smoke benchmark requires a 170MB download before it can run

### ISSUE THREAD

**Issue by johnson2006christopher** [johnson2006christopher/adaptshot#12]
## Summary

`benchmarks/run_benchmark.py --smoke-test` calls `datasets.CIFAR10(..., download=True)`
(`benchmarks/run_benchmark.py:28-31`). The documented validation gate therefore requires
an internet connection and a ~170MB download.

## Why this is a blocker, not an inconvenience

AdaptShot's stated reason to exist is that *cloud connectivity is the resource people
don't have*. The command we tell contributors to run before opening a PR does not work
without it. On a constrained connection the smoke test does not fail — it hangs:

```
$ timeout 300 python -m benchmarks.run_benchmark --smoke-test --seed 42
Terminated        (exit 124)
```

Five minutes, no output, no error. In CI that is an opaque timeout; on the connection
this library was designed for, it is a wall.

## Proposed fix

1. Vendor a small deterministic fixture (a few hundred images, or synthetic tensors with
   a fixed seed) so `--smoke-test` runs with **zero network access**.
2. Move the CIFAR-10 path behind an explicit `--dataset cifar10 --allow-download` flag.
3. Fail fast with a clear message when data is absent and downloads are not permitted.
4. Add a CI job that runs the smoke test with networking disabled, proving the claim.

## Acceptance criteria

- [ ] `python -m benchmarks.run_benchmark --smoke-test --seed 42` completes offline in < 60s
- [ ] It produces identical output on two consecutive runs with `--seed 42`
- [ ] A CI job runs it with network access blocked
- [ ] `CLAUDE.md` and `CONTRIBUTING.md` describe the offline path as the default

**Comment by johnson2006christopher** [comment]
**Correction to the original report.**

The claim above that the smoke test "does not fail — it hangs" was wrong, and the
`exit 124` evidence was misleading. That timeout happened while a concurrent first run
was still downloading CIFAR-10; the second invocation started its own download and hit
the 300s limit before either finished.

What actually happens, verified:

- **First run** (no cached data): downloads ~170MB, then completes successfully, exit 0.
- **Subsequent runs** (data cached at `data/cifar-10-batches-py`): **54.7s**, exit 0,
  determinism check passes.

```
$ /usr/bin/time -f "ELAPSED %es" python -m benchmarks.run_benchmark --smoke-test --seed 42
   • Accuracy: 68.0%
   • Avg Latency: 14.2 ms
   • P95 Latency: 15.6 ms
   • Determinism check: ✅ PASS
ELAPSED 54.71s
```

**The benchmark is not broken.** The issue is narrower than originally described, and
stands on its own terms:

1. A **one-time ~170MB download** is required before the documented validation gate can
   run at all. On a metered or intermittent connection that is a real barrier to a first
   contribution — and it is the barrier this project exists to remove.
2. A fresh CI runner pays that download **on every run**, unless the dataset is cached.
3. There is no offline path and no clear error when the download cannot proceed. It looks
   like a hang, which is exactly how it misled this report.
4. 54.7s even when cached is slow for a pre-commit gate.

The proposed fix is unchanged: ship a small vendored fixture as the default for
`--smoke-test`, put CIFAR-10 behind an explicit opt-in flag, and fail fast with a clear
message when data is absent and downloads are not permitted.

Revised acceptance criteria in the issue body.

**Comment by johnson2006christopher** [comment]
**Evidence from the first CI run of this job** (PR #37, run 32964351285):

```
Deterministic smoke benchmark ....... pass ....... 31m39s
```

**31 minutes 39 seconds** for a benchmark that takes **54.7s** locally once its data is
cached. The entire remainder is download time, paid on every push by every runner.

The job log also shows a **second** download that this issue did not originally account
for:

```
Downloading: "https://download.pytorch.org/models/resnet18-f37072fd.pth"
             to /home/runner/.cache/torch/hub/checkpoints/resnet18-f37072fd.pth
```

So the benchmark fetches two things over the network, not one:

1. **CIFAR-10** — the dataset (~170 MB), via `datasets.CIFAR10(..., download=True)`
2. **ResNet-18 pretrained weights** — the model itself, from `download.pytorch.org`

The second is the more serious of the two. A cached dataset does not help: **the backbone
weights are fetched at runtime**, so the documented validation gate cannot run on a machine
that has never been online, regardless of what data is present locally.

This widens the fix. Both must be addressed:

- Vendor a small deterministic data fixture, with CIFAR-10 behind an explicit opt-in flag
- **Bundle the backbone weights** rather than fetching them (this is what #36's ONNX work
  delivers — `pyproject.toml` already declares `package-data` for `data/*.onnx`, and
  `src/adaptshot/data/` is currently empty)
- Add CI caching for whatever must still be downloaded, so a 31-minute job does not recur

Related: #30 (prove the offline guarantee with networking hard-disabled) cannot pass while
the model weights are fetched at runtime. #36 supplies the bundled weights that make both
achievable.

**Comment by johnson2006christopher** [comment]
Done in #61, merged into `v0.3.0`.

Measured cause first: on a GitHub runner the CIFAR-10 download took **34m48s** while the benchmark itself took **56s**. 97% of every benchmark job in this repository's history was waiting for `cs.toronto.edu`, which served the runner about 80KB/s.

`--dataset` now defaults to `auto`: CIFAR-10 if cached, the synthetic fixture otherwise, and `--allow-download` must be passed explicitly. A missing dataset fails immediately with the exact command to fix it and what it costs, instead of starting a silent 170MB download.

    offline, cold:  3.4s   (was 36 minutes)
    cached CIFAR:   68.0%  unchanged

The fixture reports determinism and latency and **refuses to report accuracy** — a figure measured on random tensors describes nothing. `data_source` records provenance and a test fails if a synthetic artifact is ever committed in place of a real one.

A CI job runs the benchmark inside an empty network namespace and proves the isolation before relying on it, by opening a socket inside the namespace and failing the build if it succeeds.

**Two corrections worth recording**, both caught by that job:

1. `unshare -rn` fails on Ubuntu 24.04 runners (AppArmor restricts unprivileged user namespaces). Fixed with the documented `sysctl` workaround.
2. The offline claim was overstated. The benchmark ran offline on my machine only because `~/.cache/torch` held `resnet18-f37072fd.pth`. `core/extractor.py` fetches ImageNet weights on first use, so AdaptShot **cannot** run from a genuinely cold install with no network — that is #35/#36. Weights are now fetched in an explicit step outside the isolated one, and what the isolated step proves is stated narrowly: no dataset download, no other network call.

The job also caught a bug this PR introduced — `run_once()` in the determinism check always asked for CIFAR-10 regardless of what had been benchmarked. Fixed, with a regression test.

Also closes #60.
